naming/README.md

```diff
@@ -0,0 +1,93 @@
+
+- [Proposal] An additional instruction (or another contract) could do the following: mint an NFT for an already existing name and activate ownership for the holder of the NFT only. This would enable easy trading of names.
+
+
+Program Interface:
+
+```
+pub enum NameStatus {
+    Deleted,
+    Active,
+}
+
+pub type ValueType = u16;
+
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct NameState {
+    status: u8,
+    values: Vec<(ValueType, String)>,
+    owner_pubkey: Option<Pubkey>,
+    owner_name: Option<String>,
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
+pub enum NameInstruction {
+    /// Add a record for the given name to the naming service.
+    /// The owner can be specified as a pubkey or (exclusively) as another name (the owner then being
+    /// the owner of that name, this can be usefull for domain name extensions).
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   1. `[writable]` The program account derived from the name
+    ///   2. `[signer]` (Optional) The domain name parent account if the given name is a sub-domain
+    ///   3...N `[]` The list of accounts needed to find an owner which is a pubkey or a null
+    AddName {
+        name: String,
+        values: Vec<(ValueType, String)>,
+        owner_pubkey: Option<Pubkey>,
+        owner_name: Option<String>,
+    },
+    /// Change an entry for the given name to the naming service.
+    /// The existing value for a given Valuetype will be overwritten.
+    /// Additionnal given values will be added.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   1. `[writable]` The program account derived from the name
+    ///   2. `[signer]` The account of the first owner that is a pubkey 
+    ///   3...N `[]` The list of accounts needed to find an owner which is a pubkey or a null
+    ChangeValue {
+        name: String,
+        values: Vec<(ValueType, String)>,
+    },
+    /// Change the owner of the name.
+    /// If the owner of the given name is itself a name, the program will follow the given list of parents
+    /// until finding a pubkey owner.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   1. `[writable]` The program account derived from the name
+    ///   2. `[signer]` The current owner of the program account derived from the name
+    ///   3...N `[]` The list of accounts needed to find an owner which is a pubkey or a null
```

Comment:
what happens if this blows out the 1200-something byte transaction size limit? Very realistic if you have a long enough chain. Either we place a maximum limit on the amount of nesting allowed or you do something where you actually have to walk up the tree on the client and provide the final parent here. Question is how do you prove it's the final parent? Not sure on that one. Anyway, something to think about. There is an interesting problem to solve here that warrants more discussion.

Reply:
Yes I completely agree, it is not an easy problem. I thought that limiting the dependency depth is maybe not to worrying as Dns type of trees scale more in breadth but maybe people will come up with a new and cooler use for this. Enforcing a depth limit (of 10 parents accounts?) could be a first way of handling this, I don't see anyway of proving the parent existence without walking up the chain.. maybe we could do this in chunks and hold a temporary value in an account to not lose our progress point?

Reply:
I talked about it with @ellttBen and we agreed that adding the `parent_name` field to the NameState means that there can be an arbitrary long chain of sub-domains but all pointing to the same owner. That way having a long chain of successive owners is a border use-case that we don't need to worry about and can limit it to a maximum depth of 10?